### PR TITLE
Implementing text only parsing for Rich Text

### LINKED ISSUE #55: [Frontend?] Implement text only parsing for rich text
## 🌟 Context
This was an issue that arose from the following PR:

https://github.com/UoaWDCC/aucc/pull/53#discussion_r2060238517

For rendering rich text previews in TripReportCards and the like, we don't actually want to use a rich text renderer,  as it may render unwanted jsx such as anchor tags and images.

We want to implement a function, or custom component that can render the rich text content fetched from payload into pure text, skipping image, links, and other irrelevant types.

---

### 🎯 **Acceptance Criteria**

- [ ] A function/component is implemented that takes in a `SerializedEditorState` and converts it to text.

---

### 📊 **Definition of Done (DoD)**

- [ ] Acceptance criteria met
- [ ] Agreed tests written
- [ ] Agreed documentation in GitHub wiki
- [ ] PR reviewed

---

### 🔗 **Dependencies**

---

### PR DESCRIPTION
**Pull request by neryadom**
closes #55

## 🌟 Context 

This PR brings attention to the already existing `src/components/PlainText` directory which handles conversion of Rich Text into Plain Text.

---

## 📝 Description 

A line of description has been added to the codebase in the `src/components/PlainText/ExtractPlainText.ts` file.

---

## 📋 Notes 

The RichText global component have been left in, as we may need it down the line.

### CODE REVIEW

**Review by phawatboom — COMMENTED** [review]
Looks good! Btw, @dyzhuu If we render only plain text, shouldn't we change the content data type to text instead of a roundabout way of parsing richtext to text?

**Comment by dyzhuu** [comment]
> Looks good! Btw, @dyzhuu If we render only plain text, shouldn't we change the content data type to text instead of a roundabout way of parsing richtext to text?

@phawatboom we still want rich text, its just that when rendering previews (like the content of a trip-report in a card), we don't want some rich text elements such as images and links.

**Review by dyzhuu — CHANGES_REQUESTED** [review]
Seems like you made the changes in the other PR and it's already into main so the diff doesn't show, but I have a few comments:

It seems like these are just functions, rather than a react component, so having it under `src/components` doesn't make too much sense here. I think it makes more sense to be placed under `src/lib/utils/...` as a utility function

Also, I wonder if we really need 3 functions -- surely you can combine `getPlainText` and `extractPlainText`?

It would also be really good if you could write some tests for these functions!

**Comment by phawatboom** [comment]
> > Looks good! Btw, @dyzhuu If we render only plain text, shouldn't we change the content data type to text instead of a roundabout way of parsing richtext to text?
> 
> @phawatboom we still want rich text, its just that when rendering previews (like the content of a trip-report in a card), we don't want some rich text elements such as images and links.

Ah, I see, that makes sense. Alg then.

File: src/queries/__tests__/getPlainText.test.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  })
+
+  it('appends an ellipsis when over maxLength', () => {
+    const state = {
+      root: {
+        children: [{ type: 'text', text: 'abcdefghijklmnopqrstuvwxyz' }],
+      },
+    }
+    expect(getPlainText(state as any, 20)).toBe('abcdefghijklmnopqrst…')
+  })
```
Comment (dyzhuu) [inline src/queries/__tests__/getPlainText.test.ts:76]: This appears to  be the same test as the "truncates and appends an ellipsis when exceeding maxLength" on lines 53-60?

File: src/lib/utils/getPlainText.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *Gives you a plain-text string from a Lexical-serialized state (i.e RichText).
+ * @param data - The Lexical-serialized state to convert.
+ * @param maxLength - Optional maximum length. If provided, truncates and adds an ellipsis.
+ */
+export function getPlainText(
+  data: SerializedEditorState,
+  maxLength?: number,
+): string {
+  // recursive function to walk the node tree.
+  function walk(nodes: Node[]): string {
```
Comment (dyzhuu) [inline src/lib/utils/getPlainText.ts:19]: could we define this as another function in the same file so as to now have recursive functions?

there _probably_ shouldn't be massive performance implications if we have it like this, but best to move it just be be safe

File: src/lib/utils/getPlainText.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  // this will recursively retrieve all text from the RichText element.
+  const children = (data as any).root.children as Node[]
+  const full = walk(children).trim()
+
+  // this if statement will run if a maximum size for the plain text is passed as an arg.
+  if (typeof maxLength === 'number' && full.length > maxLength) {
+    return full.slice(0, maxLength).replace(/\s+$/, '') + '…'
+  }
+  return full
```
Comment (dyzhuu) [inline src/lib/utils/getPlainText.ts:42]:
```suggestion
  // this if statement will run if a maximum size for the plain text is passed as an arg.
  if (maxLength === undefined) return full
  
  if (full.length > maxLength) {
    return full.slice(0, maxLength).replace(/\s+$/, '') + '…'
  }
```

maybe doing an early return here is a bit cleaner? and then typescript can narrow down the type to be a number. (we could also do `if (!maxLength)` but this doesn't work for the edge case where `maxLength = 0`, depending on what you define the behavior as. 

On that thought, maybe we want to define some behaviour & tests for if maxLength= 0 and < 0?

File: src/lib/utils/getPlainText.ts
Comment (dyzhuu) [inline src/lib/utils/getPlainText.ts:1]: for utils let's stick to `kebab-case` file names, and only reserver `PascalCase` for components.

File: src/queries/__tests__/getPlainText.test.ts
Comment (dyzhuu) [inline src/queries/__tests__/getPlainText.test.ts:1]: Don't think this test belongs in the queries folder. Feel free to define a new `__tests__` folder under `lib/utils`.

**Review by dyzhuu — CHANGES_REQUESTED** [review]
Wow, this is really impressive stuff! Just a few small changes before I approve this PR :)

File: src/lib/utils/getPlainText.ts
Comment (neryadom) [reply src/lib/utils/getPlainText.ts:1]: Renamed

File: src/queries/__tests__/getPlainText.test.ts
Comment (neryadom) [reply src/queries/__tests__/getPlainText.test.ts:1]: Moved!

File: src/lib/utils/getPlainText.ts (hunk 2 above)
Comment (neryadom) [reply src/lib/utils/getPlainText.ts:19]: Took out the nested function.

File: src/queries/__tests__/getPlainText.test.ts (hunk 1 above)
Comment (neryadom) [reply src/queries/__tests__/getPlainText.test.ts:76]: Removed the test on lines 53-60. There was indeed duplication

File: src/lib/utils/getPlainText.ts (hunk 3 above)
Comment (neryadom) [reply src/lib/utils/getPlainText.ts]: Tweaked if else flow a bit

**Review by dyzhuu — APPROVED** [review]
LGTM!
